### PR TITLE
installation: Add nix flake

### PR DESCRIPTION
**Pull request by alexarice**
Makes the following two changes to make building easier on nixos:
- Adds direnv files to gitignore
- Adds a basic flake that allows `make tests` to be run

@AntonLydike

### CODE REVIEW

**Review by superlopuh — APPROVED** [review]
I know very little about nix, but if it works on your machine then I think we're safe to merge

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+  description = "XDSL devshell";
```
Comment (superlopuh) [inline flake.nix:2]: nit:
```suggestion
  description = "xDSL devshell";
```

**Comment by tobiasgrosser** [comment]
> I know very little about nix, but if it works on your machine then I think we're safe to merge

Yes, nice to see better nix support.

**Comment by alexarice** [comment]
One thing I'm unsure about is whether committing the lock file is good practice. Also might want to include `ruff` in the shell

**Comment by superlopuh** [comment]
I did a bit of googling, and it looks like it's mostly a matter of personal preference, although I found some (fixed) issues where nix used to break if the lockfile was gitignored, so at least some people thought ignoring it was a good idea. I'm happy to go either way, it doesn't seem to be a very big file.

**Review by AntonLydike — APPROVED** [review]
:partying_face:

**Comment by AntonLydike** [comment]
> One thing I'm unsure about is whether committing the lock file is good practice. Also might want to include `ruff` in the shell

I always thought committing the lock file was good practice for a project, similarly to lock files used by other package managers (notably absent from `requirements.txt` based development...)

**Review by georgebisbas — COMMENTED** [review]
Would it be an overkill to add a NixOS github actions test?https://github.com/marketplace/actions/install-nix

**Comment by AntonLydike** [comment]
> Would it be an overkill to add a NixOS github actions test?https://github.com/marketplace/actions/install-nix

Maybe we can do this in the future if we find the flake to break regularly?

**Comment by superlopuh** [comment]
My understanding is that the requirements is supposed to be a kind of lockfile, it's just not used that way consistently. The pyproj is supposed to be more flexible in terms of dependency versions, and the requirements.txt is supposed to have a fixed, known-to-work set of exact dep versions.

Anyway, I think this is good to merge, feel free to press the button at your convenience, @alexarice

**Review by georgebisbas — APPROVED** [review]

**Comment by AntonLydike** [comment]
> My understanding is that the requirements is supposed to be a kind of lockfile, it's just not used that way consistently. The pyproj is supposed to be more flexible in terms of dependency versions, and the requirements.txt is supposed to have a fixed, known-to-work set of exact dep versions.
> 
> Anyway, I think this is good to merge, feel free to press the button at your convenience, @alexarice

Notably, even `pip freeze` will not freeze versions to exact hashes, making this "lock" not as strong as e.g. poetrys lockfiles. Anyway, we are getting super off-track

**Pull request merged**